# Разбор 17 задачи (уровень: простой)

В файле содержится последовательность натуральных чисел. Элементы последовательности могут принимать целые значения от 1 до 100 000 включительно. Определите количество пар последовательности, в которых остаток от деления хотя бы одного из элементов на 117 равен минимальному элементу последовательности. В ответе запишите количество найденных пар, затем максимальную из сумм элементов таких пар. В данной задаче под парой подразумевается два идущих подряд элемента последовательности.

Файлы к заданию: https://kompege.ru/files/X2-8sfFqD.txt

## Алгоритм решения:

1) **Сначала качаем файл** в ту же директорию, где и наш файл с решением
2) **Откроем файл** с помощью команды `open` и прочтём данные с помощью метода `readlines()`. Почему именно `readlines()`? Внутри файла числа идут подряд с новой строки. Поэтому мы хотим прочитать все строки, что есть в файле. В результате мы получим массив со строчными объектами.

In [22]:
file = open("17_4597.txt").readlines()
print(file)

['567\n', '51360\n', '76956\n', '45215\n', '22129\n', '34648\n', '34651\n', '69023\n', '56338\n', '37358\n', '26733\n', '8515\n', '55037\n', '85160\n', '22214\n', '13563\n', '62844\n', '18213\n', '48113\n', '68663\n', '4760\n', '33821\n', '64418\n', '3269\n', '1723\n', '76862\n', '27814\n', '84747\n', '13688\n', '42978\n', '26437\n', '87723\n', '32354\n', '72268\n', '78795\n', '87110\n', '13336\n', '14698\n', '35676\n', '44273\n', '20702\n', '37866\n', '20044\n', '40295\n', '61990\n', '66101\n', '79858\n', '23895\n', '46221\n', '14898\n', '23404\n', '5022\n', '49918\n', '85452\n', '47063\n', '8584\n', '852\n', '42715\n', '49574\n', '16075\n', '9056\n', '61997\n', '40688\n', '40641\n', '78726\n', '44470\n', '7318\n', '83474\n', '80200\n', '9927\n', '4682\n', '40058\n', '45746\n', '61215\n', '14070\n', '27354\n', '11877\n', '65741\n', '26009\n', '62451\n', '81080\n', '53141\n', '10803\n', '77333\n', '9240\n', '52549\n', '41811\n', '6608\n', '58419\n', '15968\n', '61243\n', '79879\n', '57

Обратим наше внимание на 3 предложение задачи:

**"Определите количество пар последовательности, в которых остаток от деления хотя бы одного из элементов на 117 равен минимальному элементу последовательности."**

Очень много задач, которые нужно реализовать в одном предложении. Давайте выпишем все задачи, которые нам нужны:

1) Определить количество пар в последовательности
2) В которых хотя бы одно число при делении на 117 даёт остаток
3) Остаток, который равен минимальному элементу последовательности

**Для того чтобы ответить на все 3 вопроса, нам нужно пойти в обратном порядке:**

1) Сначала найти минимальное число последовательности
2) Создать пары элементов  
3) Проверить эти пары на то, что хотя бы одно из них при делении на 117 даёт остаток, равный минимальному числу, которое мы найдём в 1 пункте

**Важный подготовительный этап:**

Перед выполнением всех пунктов нам необходимо перевести все строчные объекты в списке `file` в числовые объекты.

In [21]:
file = list(map(int,file))
print(file)

[567, 51360, 76956, 45215, 22129, 34648, 34651, 69023, 56338, 37358, 26733, 8515, 55037, 85160, 22214, 13563, 62844, 18213, 48113, 68663, 4760, 33821, 64418, 3269, 1723, 76862, 27814, 84747, 13688, 42978, 26437, 87723, 32354, 72268, 78795, 87110, 13336, 14698, 35676, 44273, 20702, 37866, 20044, 40295, 61990, 66101, 79858, 23895, 46221, 14898, 23404, 5022, 49918, 85452, 47063, 8584, 852, 42715, 49574, 16075, 9056, 61997, 40688, 40641, 78726, 44470, 7318, 83474, 80200, 9927, 4682, 40058, 45746, 61215, 14070, 27354, 11877, 65741, 26009, 62451, 81080, 53141, 10803, 77333, 9240, 52549, 41811, 6608, 58419, 15968, 61243, 79879, 57065, 84841, 1030, 84471, 23236, 55815, 53154, 52676, 41351, 79461, 79650, 61406, 23113, 58123, 32278, 30767, 66758, 82942, 85684, 3367, 81654, 6456, 50639, 57549, 78772, 41933, 26873, 10297, 24817, 30089, 50386, 2194, 13806, 66814, 30908, 23493, 88425, 40264, 82109, 22692, 81181, 28966, 67900, 63405, 7625, 6072, 59207, 40818, 65938, 73461, 59319, 79046, 39076, 81909,

Теперь, когда мы имеем список с числовыми объектами, перейдём к выполнению всех пунктов.

**Пункт 1:**

In [6]:
mn = min(file)
print(mn)

8


Минимальное число в последовательности: 8

**2 пункт:** Создадим пары элементов. Пары элементов можно составить разными комбинациями. Поэтому обратимся к последнему предложению задачи.

Там сказано: **"В данной задаче под парой подразумевается два идущих подряд элемента последовательности."** Реализуем данные пары с помощью индексов списка.

In [24]:
for i in range(len(file)-1):  # Цикл по индексам от 0 до предпоследнего элемента
    num1 = file[i]           # Получаем текущий элемент последовательности
    num2 = file[i+1]         # Получаем следующий элемент последовательности
    print(num1,num2)         # Выводим пару соседних элементов

567
 51360

51360
 76956

76956
 45215

45215
 22129

22129
 34648

34648
 34651

34651
 69023

69023
 56338

56338
 37358

37358
 26733

26733
 8515

8515
 55037

55037
 85160

85160
 22214

22214
 13563

13563
 62844

62844
 18213

18213
 48113

48113
 68663

68663
 4760

4760
 33821

33821
 64418

64418
 3269

3269
 1723

1723
 76862

76862
 27814

27814
 84747

84747
 13688

13688
 42978

42978
 26437

26437
 87723

87723
 32354

32354
 72268

72268
 78795

78795
 87110

87110
 13336

13336
 14698

14698
 35676

35676
 44273

44273
 20702

20702
 37866

37866
 20044

20044
 40295

40295
 61990

61990
 66101

66101
 79858

79858
 23895

23895
 46221

46221
 14898

14898
 23404

23404
 5022

5022
 49918

49918
 85452

85452
 47063

47063
 8584

8584
 852

852
 42715

42715
 49574

49574
 16075

16075
 9056

9056
 61997

61997
 40688

40688
 40641

40641
 78726

78726
 44470

44470
 7318

7318
 83474

83474
 80200

80200
 9927

9927
 4682

4682
 40058

40058
 45746

45746
 61215

6121

`for i in range(len(file)-1):` - цикл проходит по всем индексам от 0 до длины списка минус 2. Минус 1 нужен потому, что мы обращаемся к элементу i+1, иначе на последней итерации получим выход за границы списка.

`num1 = file[i]` - берём элемент с текущим индексом i. Например, если file = [10, 20, 30, 40], то при i=0 получим num1=10.

`num2 = file[i+1]` - берём следующий элемент. Для того же примера при i=0 получим num2=20.

**3 пункт:** Проверим на деление

Проверим условие, что хотя бы одно из чисел пары при делении на 117 даёт остаток, равный минимальному элементу последовательности (8).

In [14]:
c = 0  # Инициализация счетчика пар, удовлетворяющих условию
for i in range(len(file)-1):  # Цикл по индексам от 0 до предпоследнего элемента
    num1 = file[i]           # Получаем текущий элемент последовательности
    num2 = file[i+1]         # Получаем следующий элемент последовательности
    if num1 % 117 == mn or num2 % 117 == mn:  # Проверяем условие для хотя бы одного из двух чисел
        c += 1               # Увеличиваем счетчик, если условие выполняется
print(c)  # Выводим общее количество пар, удовлетворяющих условию

175


`num1 % 117 == mn` - проверяет, даёт ли текущее число при делении на 117 остаток, равный значению переменной `mn`

`num2 % 117 == mn` - аналогичная проверка для следующего числа

`or` - условие выполняется, если **ХОТЯ БЫ ОДНО** из двух чисел удовлетворяет критерию

Отлично, первый ответ мы получили, но в задаче от нас просят 2 числа.

Посмотрим на вопрос задачи: **"В ответе запишите количество найденных пар, затем максимальную из сумм элементов таких пар."**

Для того чтобы найти максимальную сумму элементов подходящих пар, нам нужна переменная, в которую мы будем запоминать максимальную сумму.

In [16]:
mx = 0  # Инициализация переменной для хранения максимальной суммы пары
c = 0  # Инициализация счетчика пар, удовлетворяющих условию
for i in range(len(file)-1):  # Цикл по индексам от 0 до предпоследнего элемента
    num1 = file[i]           # Получаем текущий элемент последовательности
    num2 = file[i+1]         # Получаем следующий элемент последовательности
    if num1 % 117 == mn or num2 % 117 == mn:  # Проверяем условие для хотя бы одного из двух чисел
        c += 1               # Увеличиваем счетчик, если условие выполняется
        mx = max(mx, num1+num2)  # Обновляем максимальную сумму, если текущая сумма больше предыдущего максимума
print(c, mx)  # Выводим количество подходящих пар и максимальную сумму среди них

175 173738


`mx = 0` - инициализация переменной для хранения максимальной суммы пары (начинаем с 0, так как суммы чисел всегда ≥ 0)

`mx = max(mx, num1+num2)` - функция `max()` сравнивает текущее значение `mx` с суммой текущей пары `num1+num2` и сохраняет большее значение обратно в `mx`

В итоге мы получаем максимальную сумму пар: **173738**

**Итоговые ответы в нашей задаче:** 175 173738

**Итоговый код:**

In [25]:
# Открываем файл "17_4597.txt" и читаем все строки в список
file = open("17_4597.txt").readlines()

# Преобразуем все строки в список целых чисел
file = list(map(int,file))

# Находим минимальное значение в списке чисел
mn = min(file)

mx = 0  # Инициализация переменной для хранения максимальной суммы пары
c = 0  # Инициализация счетчика пар, удовлетворяющих условию

for i in range(len(file)-1):  # Цикл по индексам от 0 до предпоследнего элемента
    num1 = file[i]           # Получаем текущий элемент последовательности
    num2 = file[i+1]         # Получаем следующий элемент последовательности
    
    # Проверяем условие: хотя бы одно число должно давать остаток mn при делении на 117
    if num1 % 117 == mn or num2 % 117 == mn:
        c += 1               # Увеличиваем счетчик, если условие выполняется
        mx = max(mx, num1+num2)  # Обновляем максимальную сумму пар

# Выводим количество подходящих пар и максимальную сумму среди них
print(c, mx)

175 173738
